### PR TITLE
test(unify): Specs Page Workflows - Component Tests

### LINKED ISSUE #19207: Unification App: Write E2E tests around "Specs Page" - New component testing project
### What would you like?

Write end-to-end tests to cover the new Unification work in 10.0-release branch for "[Choose a Browser](https://docs.google.com/spreadsheets/d/1iPwi89aW6aYeA0VT1XOhYdAWLuScW0okrlfcL9fzh3s/edit#gid=0)" in the App around what the specs page when on a new component testing project. 

### Why is this needed?

_No response_

### Other

_No response_

### PR DESCRIPTION
**Pull request by tbiethman**
- Closes #19207 

### User facing changelog

N/A

### Additional details

The following reqs are accounted for here:

New Project - Specs Page (component)
* When there are no specs detected (and user has not overridden `defaultSpecPattern`), the page has selections for 'Create from story' and 'Create from component'
* When there are no specs detected, there's a button to 'View spec pattern'
* Clicking the 'View spec pattern' button will redirect you to the Settings page
* Clicking 'Create from story' will open a 'Choose a story' modal
* Clicking 'Create from component' will open a 'Choose a component' modal

New Project - Specs Page - Choose a Story (component)
* The modal can be closed by clicking an 'x' button and take you back to specs empty page
* The 'modal has a default file extension filter of *.stories
* The modal has a file extension button that can be clicked to type in the file extension. This will filter the results by file extension found
* The number of files matching from the search will display in the searchbar
* Clicking on a matching story will route you to a 'Spec successfully added' modal

New Project - Specs Page - Choose a component (component)
* The modal can be closed by clicking an 'x' button and take you back to specs empty page
* There's a default file extension filter of *.(component testing framework)
* There's a file extension button that can be clicked to type in the file extension. This will filter the results by file extension found
* The number of files matching from the search will display in the searchbar
* Hovering over a matching file will display its relative path
  * TODO - needs UI enhancement to make file list items keyboard accessible
* Clicking on a matching story will route you to a 'Spec successfully added' modal

New Project - Specs Page - Spec Successfully Added (component)
* The 'Spec successfully added' modal will display the spec's relative path that was added in green
* The new specfile has been created in your file system
* Clicking the spec's relative path will expand to display a preview of the specfile's contents
* Clicking 'x' to close the modal will display the Specs list page with the current spec
* Clicking 'create another spec' will display the 'Create from story' and 'Create from component' options within the modal
* Clicking 'Okay, run the spec' button will open the spec in the Test Runner

### PR Tasks

- [X] Have tests been added/updated?
- [X] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [n/a] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [n/a] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
- [n/a] Have new configuration options been added to the [`cypress.schema.json`](https://github.com/cypress-io/cypress/blob/develop/cli/schema/cypress.schema.json)?

### CODE REVIEW

File: packages/frontend-shared/src/components/Card.vue [hunk 1]
```diff
@@ -8,7 +8,7 @@
       'bg-gray-50 border-gray-100 pointer-events-none': disabled
     }"
     data-cy="card"
-    :tabindex="disabled ? -1 : 1"
+    :tabindex="disabled ? -1 : 0"
```
Comment (tbiethman) [inline packages/frontend-shared/src/components/Card.vue:11]: The tabindex of 1 pulled the cards out of the natural tab order. Talked with @marktnoonan, the cards are going to get a larger markup change in the future.

File: packages/frontend-shared/src/components/StandardModalHeader.vue [hunk 2]
```diff
@@ -2,8 +2,11 @@
   <div
     class="sticky top-0 flex items-center justify-between bg-white rounded-t border-b-1px border-b-gray-100 min-h-56px px-24px z-1"
   >
-    <DialogTitle class="text-gray-900 text-18px">
-      <slot /> <span
+    <div>
```
Comment (tbiethman) [inline packages/frontend-shared/src/components/StandardModalHeader.vue:5]: Adding an additional container here to remove the 'Need help?' link from the dialog's label while keeping the desired styling. Prior to this change, the dialog would read as "Choose a spec Need help?"; now it's just "Choose a spec".

File: packages/app/cypress/e2e/specs_page.cy.ts [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+            // the storybook card remains disabled here
+            cy.findAllByTestId('card').eq(1)
+            .should('have.attr', 'tabindex', '-1')
+          })
+        })
+      })
+    })
+  })
+
+  describe('Code Generation', () => {
```
Comment (tbiethman) [inline packages/app/cypress/e2e/specs_page.cy.ts:282]: Full credit to @ZachJW34 for writing the Code Generation tests, I moved them into into this file to co-locate them with the rest of the specs page tests.

**Review by marktnoonan — CHANGES_REQUESTED** [review]
Looks good! Just a couple of suggestions in the tests. 

And then, not for this PR, but: these tests also makes me think we could refactor our tabindex checks into a custom chai assertion so that we could use `be.focusable` and `not.be.focusable`, it would be good to talk about that . There's a [recipe](https://github.com/cypress-io/cypress-example-recipes/blob/master/examples/extending-cypress__chai-assertions/cypress/support/index.js) that looks pretty straightforward.

**Comment by tbiethman** [comment]
> And then, not for this PR, but: these tests also makes me think we could refactor our tabindex checks into a custom chai assertion so that we could use be.focusable and not.be.focusable, it would be good to talk about that . There's a recipe that looks pretty straightforward.

That's a good point. Even a custom cy command to call `focus` on an element and assert that is it actually focused afterwards might be helpful, if a little more intrusive to the AUT. Ideally, we wouldn't be explicitly checking tabindex at all and relying on 'disabled' attributes where possible; the Card refactor will help a lot there.

File: packages/app/cypress/e2e/specs_page.cy.ts [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+        })
+      })
+
+      it('shows create first spec page with create from component and create from story options', () => {
+        cy.findByRole('heading', {
+          level: 1,
+          name: defaultMessages.createSpec.page.defaultPatternNoSpecs.title,
+        }).should('be.visible')
+
+        cy.findByTestId('create-spec-page-description').should('be.visible').and('contain', defaultMessages.createSpec.page.defaultPatternNoSpecs.component.description)
```
Comment (emilyrohrbough) [inline packages/app/cypress/e2e/specs_page.cy.ts:38]:
```suggestion
        cy.findByTestId('create-spec-page-description')
        .should('be.visible')
        .and('contain', defaultMessages.createSpec.page.defaultPatternNoSpecs.component.description)
```

File: packages/app/cypress/e2e/specs_page.cy.ts [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+
+        cy.get('@ComponentCard').should('be.visible')
+        cy.get('@StoryCard').should('be.visible')
+
+        cy.findByTestId('no-specs-message').should('be.visible').and('contain', defaultMessages.createSpec.noSpecsMessage)
```
Comment (emilyrohrbough) [inline packages/app/cypress/e2e/specs_page.cy.ts:43]:
```suggestion
        cy.findByTestId('no-specs-message')
        .should('be.visible')
        .and('contain', defaultMessages.createSpec.noSpecsMessage)
```

**Review by marktnoonan — DISMISSED** [review]

File: packages/app/cypress/e2e/specs_page.cy.ts [hunk 6]
```diff
@@ -0,0 +1,353 @@
+import defaultMessages from '@packages/frontend-shared/src/locales/en-US.json'
+
+describe('Specs Page', () => {
+  describe('Component Workflows', {
+    viewportHeight: 768,
+    viewportWidth: 1024,
+  }, () => {
+    context('with storybook', () => {
```
Comment (emilyrohrbough) [inline packages/app/cypress/e2e/specs_page.cy.ts:8]:
```suggestion
    context('create spec with storybook', () => {
```

Comment (emilyrohrbough) [reply packages/app/cypress/e2e/specs_page.cy.ts:8]: maybe
`project with storybook`

File: packages/app/cypress/e2e/specs_page.cy.ts [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+
+            // the storybook card remains enabled here
+            cy.findAllByTestId('card').eq(1)
+            .should('have.attr', 'tabindex', '0')
+          })
+        })
+      })
+    })
+
+    context('without storybook', () => {
```
Comment (emilyrohrbough) [inline packages/app/cypress/e2e/specs_page.cy.ts:145]:
```suggestion
    context('project without storybook', () => {
```

File: packages/app/cypress/e2e/specs_page.cy.ts [hunk 8]
```diff
@@ -0,0 +1,353 @@
...
+
+        cy.findByTestId('no-specs-message').should('be.visible').and('contain', defaultMessages.createSpec.noSpecsMessage)
+
+        cy.findByRole('link', { name: defaultMessages.createSpec.viewSpecPatternButton })
+        .should('be.visible')
+        .and('not.be.disabled')
+        .and('have.attr', 'href', '#/settings?section=project&setting=specPattern')
+      })
+
+      context('create from story', () => {
```
Comment (emilyrohrbough) [inline packages/app/cypress/e2e/specs_page.cy.ts:51]: Would it be worth throwing a test in to verify that clicking create from component correctly finds matches?

It does look like the stories are matched as well which makes it seem like we won't even need the create from story option...
Screen Shot 2022-01-13 at 2 35 32 PM

File: packages/app/cypress/e2e/specs_page.cy.ts [hunk 9]
```diff
@@ -0,0 +1,353 @@
...
+      cy.startAppServer('component')
+      cy.visitApp()
+    })
+
+    const checkCodeGenCandidates = (specs: string[]) => {
+      cy.findByTestId('file-match-indicator').contains(`${specs.length} Match${specs.length > 1 ? 'es' : ''}`)
+      cy.findAllByTestId('file-list-row').should('have.length', specs.length)
+      .each((row, i) => cy.wrap(row).contains(specs[i]))
+    }
+
```
Comment (emilyrohrbough) [inline packages/app/cypress/e2e/specs_page.cy.ts:297]: It'd be worth beefing this up to test for:
- generating spec from component/story multiple times --  naming and code generation
- cannot generate spec from a spec
- generating a spec from a story when the `create from component` route is used

File: packages/app/cypress/e2e/top-nav.cy.ts [hunk 10]
```diff
@@ -194,11 +194,12 @@ describe('App Top Nav Workflows', () => {
 
         cy.findByTestId('cypress-update-popover').findByRole('button', { name: 'Update to 10.1.0' }).click()
 
-        cy.findByRole('dialog', { name: 'Upgrade to Cypress 10.1.0 Need help?' }).as('upgradeModal')
-        .should('contain', 'You are currently running Version 10.0.0 of Cypress')
-        .and('contain', 'npm install --save-dev cypress@10.1.0')
-
-        cy.get('@upgradeModal').findByRole('button', { name: 'Close' }).click()
+        cy.findByRole('dialog', { name: 'Upgrade to Cypress 10.1.0' }).as('upgradeModal').within(() => {
```
Comment (emilyrohrbough) [inline packages/app/cypress/e2e/top-nav.cy.ts:197]: are 10.0.0 and 10.1.0 hard coded?

**Review by emilyrohrbough — DISMISSED** [review]
Nice job. Overall it seems like this is pretty much there. My bigger questions are around the implementation, not the tests themselves.

File: packages/app/cypress/e2e/top-nav.cy.ts (hunk 10 above)
Comment (emilyrohrbough) [reply packages/app/cypress/e2e/top-nav.cy.ts:197]: ahh I should have expanded these further to view the full spec

**Comment by tbiethman** [comment]
A lot has changed since I've been pulled away from these. Going to close this in favor of a new PR: https://github.com/cypress-io/cypress/pull/19780

**Pull request closed without merging**